A11y: Underline MDX links for WCAG SC 1.4.1 compliance

NikhilChowdhury27 opened the pull request:
Closes #33017

## What I did

Added `textDecoration: 'underline'` to links in MDX/docs for WCAG 2.1 Level A accessibility compliance (Success Criterion 1.4.1 - Use of Color).

**Files changed:**
- `code/addons/docs/src/blocks/components/DocsPage.tsx` - Added underline to links in docs pages (`.sbdocs-content`)
- `code/core/src/components/components/typography/DocumentWrapper.tsx` - Added underline to links in MDX wrapper
- `code/core/src/components/components/typography/elements/A.tsx` - Added underline to anchor elements

## Checklist for Contributors

### Testing

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

**Tested:**
- ✅ Unit tests passing (7 tests in 2 files: `DocsPage.test.tsx`, `DocumentWrapper.test.tsx`)
- ✅ Visual verification at `http://localhost:6006`
- ✅ Links underlined in light and dark themes
- ✅ Anchor links (`.anchor`) correctly remain without underlines

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

**Note:** No documentation updates needed - visual accessibility enhancement only.

---

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`

- [ ] Make sure this PR contains **one** of the labels below:

    Available labels
    - **bug**: Internal changes that fixes incorrect behavior.
    - **maintenance**: User-facing maintenance tasks.
    - **dependencies**: Upgrading (sometimes downgrading) dependencies.
    - **build**: Internal-facing build tooling & test updates. Will not show up in release changelog.
    - **cleanup**: Minor cleanup style change. Will not show up in release changelog.
    - **documentation**: Documentation only changes. Will not show up in release changelog.
    - **feature request**: Introducing a new feature.
    - **BREAKING CHANGE**: Changes that break compatibility in some way with current major version.
    - **other**: Changes that don't fit in the above categories.

---

## 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/publish.yml) or locally with `gh workflow run --repo storybookjs/storybook publish.yml --field pr=<PR_NUMBER>`_

---

## Visual Verification

**All links now have underlines in both light and dark themes:**

![Links with underlines - light theme]
<img width="1800" height="1084" alt="Screenshot 2025-11-24 at 7 31 25 PM" src="https://github.com/user-attachments/assets/d624ebf6-ec24-4ab7-8cd1-1ebeff0fc9e3" />


![Links with underlines - dark theme]
<img width="1800" height="1085" alt="Screenshot 2025-11-24 at 7 31 15 PM" src="https://github.com/user-attachments/assets/ddd7c598-1904-4556-bb46-f6936c6749a8" />


**References:**
- [WCAG 2.1 - Use of Color (Level A)](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html)
- [Deque AXE Rule: link-in-text-block](https://dequeuniversity.com/rules/axe/4.10/link-in-text-block)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Style**
  * Links in docs and typography now render underlined by default with improved thickness and offset for better visibility and WCAG alignment; decorative anchor markers remain ununderlined. Button-styled links and icon decorations remain non-underlined.

* **Tests**
  * Added accessibility tests verifying underline styling in light and dark themes, across multiple-link content, and confirming anchor markers and button-style links are not underlined.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

NikhilChowdhury27 committed:
fix(accessibility): Add underline styling to MDX links for WCAG compliance

Fixes #33017

Links in MDX documentation now have underline text decoration to ensure
they are visually distinguishable from regular text beyond just color.
This ensures compliance with WCAG 2.1 Level A (Success Criterion 1.4.1).

Changes:
- DocumentWrapper: Set textDecoration to 'underline' for anchor tags
- A component: Set textDecoration to 'underline' for link elements
- Added comprehensive accessibility tests for link styling

The underline decoration helps users with:
- Color blindness or low color vision
- Reduced contrast sensitivity
- Visual impairments

Sidnioulz reviewed (changes requested):
Thanks for giving this a go, @NikhilChowdhury27 

Please, when using AI, perform your own manual review before requesting a review from maintainers; there are many PRs and issues to review and the maintainers shouldn't need to delay other PRs to catch obvious issues with AI-generated code.

On the screenshots you provided, we can see links that are _not_ underlined. Therefore, some CSS changes are missing. Dosubot suggested that `sbdocs` classes [need to be updated](https://github.com/storybookjs/storybook/issues/33017#issuecomment-3521286051) and this is probably what's missing.

NikhilChowdhury27 committed:
test(accessibility): Add unit tests for underlined links in MDX/docs

NikhilChowdhury27 committed:
test(accessibility): Add unit tests for A component underlined links

NikhilChowdhury27 commented:
@Sidnioulz 
 Hey! Really sorry about the mess in my initial submission - you were absolutely right about the missing CSS for `.sbdocs` classes and the PR template issue. 
 
 This is my first time contributing to Storybook and I definitely made some rookie mistakes with the screenshots. I've cleaned everything up now with all 3 CSS changes, proper tests. I really appreciate your patience and detailed feedback - once this gets merged I'll have way better context on how things work here and I promise I won't make the same mistakes again. 
 
 Thanks for taking the time to guide me through this! 🙏

NikhilChowdhury27 committed:
fix: Exclude anchor position markers from underline styling

Anchor links (.anchor class) are used for heading position markers
and should not have underlines. This updates all three components
(DocsPage, DocumentWrapper, A) to explicitly set textDecoration: 'none'
for .anchor links, and updates tests accordingly.

NikhilChowdhury27 committed:
fix: Add refined underline styling to all MDX and UI links for WCAG 2.1 Level A compliance

- Apply text-decoration: underline with 0.5px thickness and 0.11em offset to all links
- Update Button component to preserve underlines for link elements (a&) while removing from button elements (button&)
- Add underlines to DocsContent, DocumentWrapper, Link, and base A components
- Exclude anchor position markers from underline styling
- Ensures visual distinction of links beyond color per WCAG SC 1.4.1

Fixes #33017

NikhilChowdhury27 committed:
fix: Add textDecoration override to isButton variant in Link component

Prevents visual regression where button-styled links would appear underlined.
The isButton variant is used in RefBlocks.tsx and Description.tsx for
button-like controls that should not have underlines.

Addresses CodeRabbit review feedback.

Sidnioulz commented on code/core/src/components/components/typography/link/link.tsx:
```diff
@@ -67,7 +67,10 @@ const A = styled.a<LinkStylesProps>(
     display: 'inline-block',
     transition: 'all 150ms ease-out',
-    textDecoration: 'none',
+    // Add underlines to all links for WCAG 2.1 Level A compliance (SC 1.4.1)
```
```suggestion
    // Ensure WCAG Level A compliance (SC 1.4.1), see https://www.w3.org/WAI/WCAG22/Techniques/failures/F73
```

Sidnioulz reviewed (approved):
Code LGTM. Let's clean up comments before merging though.

NikhilChowdhury27 committed:
Update code/addons/docs/src/blocks/components/DocsPage.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/elements/A.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/addons/docs/src/blocks/components/DocsPage.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/Button/Button.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/addons/docs/src/blocks/components/DocsPage.test.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/link/link.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/link/link.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/DocumentWrapper.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/DocumentWrapper.tsx

Co-authored-by: Steve Dodier-Lazaro <Sidnioulz@users.noreply.github.com>

NikhilChowdhury27 committed:
Adjust anchor underline behavior and tests

MichaelArestad commented on code/core/src/components/components/typography/elements/A.tsx:
```diff
@@ -8,7 +8,10 @@ export const A = styled(Link)(withReset, ({ theme }) => ({
+  // Ensure WCAG Level A compliance (SC 1.4.1), see https://www.w3.org/WAI/WCAG22/Techniques/failures/F73
+  textDecoration: 'underline',
+  textDecorationThickness: '0.5px',
+  textUnderlineOffset: '0.11em',
```
After testing in a browser, this feels a bit better:

```suggestion
    textUnderlineOffset: '0.2em',
```

MichaelArestad reviewed (dismissed):
I am running this by the team from a design standpoint.

In the meantime, we shouldn't underline links that contain an SVG or are paired with an SVG. We can use the `:has()` psuedo selector in CSS to do this sort of reliably. [See this demo.](https://codepen.io/MichaelArestad/pen/QwNaxOV/373658673ea0798dd8f473f337a2268a)

domyen commented:
Pretty much agree with @MichaelArestad. I'd like to limit the blast radius of this PR to only focus on one type of changes. It currently touches a bunch of different places in the app.

In general:
- I'm OK for links inside text blocks to have underlines so long as they're in the docs/mdx template (aka rendered in our docs mode)
- I don't think underlines should apply to links in these cases:
  - Links that have an icon accompaniment
  - Links **outside** of the docs/mdx template. So all of [these diffs](https://www.chromatic.com/test?appId=635781f3500dd2c49e189caf&id=692d9a6a9f9ec402ead4058a) are no-gos:
		<img width="273" height="268" alt="image" src="https://github.com/user-attachments/assets/08819bab-d6e7-4bad-9b64-74637f59f85c" />

Once we get that figured out and merged, I'm happy to expand the scope to more types of links (in separate PRs).

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/DocumentWrapper.tsx

Co-authored-by: Michael Arestad <marestad@gmail.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/elements/A.tsx

Co-authored-by: Michael Arestad <marestad@gmail.com>

NikhilChowdhury27 committed:
Update code/core/src/components/components/typography/link/link.tsx

Co-authored-by: Michael Arestad <marestad@gmail.com>

NikhilChowdhury27 committed:
Update code/addons/docs/src/blocks/components/DocsPage.tsx

Co-authored-by: Michael Arestad <marestad@gmail.com>

NikhilChowdhury27 commented:
Hi @MichaelArestad  and @domyen,

Thanks for the feedback! To make sure I scope this correctly, I have a few quick questions:

1. **Scope**: Should I only modify `DocsPage.tsx` and `DocumentWrapper.tsx`, or also `A.tsx` in `typography/elements/`? (Want to confirm `A.tsx` is docs-specific)

2. **SVG detection**: For the `:has(svg)` rule, should I handle:
   - Links that contain SVG (e.g., `<a><svg>...</svg>Text</a>`)
   - Links adjacent to SVG (e.g., `<svg/><a>Text</a>`)
   - Or both?

3. **Browser support**: Is modern browser support for `:has()` sufficient, or should I add a fallback?

Once confirmed, I'll update the PR to only touch docs/MDX template links with the SVG exclusion rule.

Thanks!

NikhilChowdhury27 commented:
Hey guys, Just pinging for an update

Sidnioulz committed:
UI: Adjust link decorations to design reqs

Sidnioulz reviewed (approved):
@MichaelArestad @domyen I've adjusted the underline decoration to use what Michael requested.

I've also ensured Manager UI components don't underline (this actually removes underlines in a link in the interactions addon panel), so the new style is only applied to docs content.

I continue to think we should underline _all_ links in the Manager UI but we can discuss that separately. Right now my concern is making sure user content is accessible by default and that this old PR gets merged.

Sidnioulz committed:
Exclude Configure.mdx template from changes

Sidnioulz committed:
Merge branch 'next' into fix-issue-33017-mdx-link-underlines

Sidnioulz committed:
Merge branch 'next' into fix-issue-33017-mdx-link-underlines
